### DISCUSSION TITLE
sbt-heroku plugin no longer maintained

### DISCUSSION THREAD

**Discussion by sief** [playframework/playframework#12334] [General]
The sbt-heroku plugin as described here https://www.playframework.com/documentation/3.0.x/ProductionHeroku#Deploying-with-the-sbt-heroku-plugin is no longer maintained: https://github.com/heroku/heroku-sbt-plugin

To deploy directly to Heroku without pushing to a Git repository, you need to use the heroku-jvm-application-deployer https://github.com/heroku/heroku-jvm-application-deployer which requires a fat Jar. 

The Play documentation describes creating a fat jar using the sbt-assembly plugin as not officially supported: https://www.playframework.com/documentation/3.0.x/Deploying#Using-the-sbt-assembly-plugin

What is the reason that it's not officially supported?

**Comment by mkurz** [comment]
> What is the reason that it's not officially supported?

It's been a while since this docs were added in
- #4194

I guess the meaning was that the sbt-assembly plugin (https://github.com/sbt/sbt-assembly) was not / is not maintained by the Playframework maintainers. So if you use it and run into problems you are on your own or have to contact the author of the plugin. However, meanwhile, I am pretty sure the plugin should be safe to use, it's quite popular.

@ndeverge So you forked the heroku-sbt-plugin, any reason you can not migrate to the Heroku JVM Application Deployer?

**Comment by ndeverge** [comment]
I tried to create a fat jar to use with the heroku-jvm-application-deployer, but it did not work (I did not have the time to investigate further at the moment), and it was easier for me to fork the sbt-heroku plugin.

**Comment by sief** [comment]
Same here. The sbt-native-packager (stage or dist) works out of the box, as does the sbt-heroku plugin when used with `sbt stage`. I had never used the sbt-assembly plugin before and when I tried it with my multi project sbt build, I got a ton of deduplicate errors. Is there a way to build a fat Jar without adjusting these merge strategies, similar to what the  sbt-native-packager does?

**Comment by mkurz** [comment]
Sorry I don't have time to help out here, since this is not a direct Play problem. You probably want to take a look what the `deployHeroku` of the archived plugin actually does and somehow convert that to make that happen in sbt-assembly:
* https://github.com/heroku/heroku-sbt-plugin/blob/main/src/main/scala/com/heroku/sbt/HerokuPlugin.scala#L45-L53
* https://github.com/heroku/heroku-sbt-plugin/blob/main/src/main/scala/com/heroku/sbt/SbtApp.scala#L57

Or just keep using the forked plugin from @ndeverge for now.

If you figure out how to configure sbt-assembly for Heroku it would be good if we could update [our docs](https://www.playframework.com/documentation/3.0.x/ProductionHeroku#Deploying-to-Heroku). Pull requests welcome.

**Reply by mkurz** [reply]
And by the way, you don't even need to fork the plugin to make use of it (If it still works currently of course). Instead you can use:
```sbt
lazy val herokuSbtPlugin = RootProject(uri("https://github.com/heroku/heroku-sbt-plugin.git#c058dd2593611335d8771b3ab01494b2b5c6855d"))
lazy val root = project.in(file(".")).dependsOn(herokuSbtPlugin)
```
in your `project/plugins.sbt` which directly references the github repo pointing to the latest commit.
